Added support for string parameters and variables

LuigiPulcini opened the pull request:
If there is any interest in adding support for string parameters and variables, this PR adjusts the version forked by @issidorov (https://github.com/issidorov/fparse/), updating it to the latest master branch (including the refactoring to Typescript).

This is also requested in #51.

An example of what this version can do is offered below:

```javascript
const f1 = new Formula( 'if( compare( [quantity], 10, ">=" ), .9, 1 ) * [price]' );

f1.if = ( value, ifTrue, ifFalse ) => {
    return ! ! value ? ifTrue : ifFalse;
};

f1.compare = ( value, reference, comparison ) => {
    switch ( comparison ):
        case '>=':
            return value >= reference ? 1 : 0;
       ...
};

const res = f1.evaluate( { quantity: 10, price: 100 } );
console.log( res );

// res = 900
```

I hope this helps.

bylexus reviewed (commented):
Thanks for your contribution. I started the review by checking out your branch, building it (using nodejs 20), and ran the tests (npm run test). There are 13 tests failing.

Would you have a look into it, or maybe I did oversee something?
Thanks

bylexus commented on spec/specs/stringBasicSpec.js:
```diff
@@ -0,0 +1,96 @@
+    });
+
+    it('correct in parentheses', function () {
+        expect(Fparser.calc('(")")')).toEqual(')');
```
this test fails.

bylexus commented on spec/specs/stringBasicSpec.js:
```diff
@@ -0,0 +1,96 @@
+    });
+
+    it('support usage by variable', function () {
+        expect(Fparser.calc('x', { x: 'foo' })).toEqual('foo');
```
this test fails.

bylexus commented on spec/specs/stringBasicSpec.js:
```diff
@@ -0,0 +1,96 @@
+        });
+    });
+
+    it('blocking math operator "-"', function () {
```
this test fails.

bylexus commented on spec/specs/stringBasicSpec.js:
```diff
@@ -0,0 +1,96 @@
+        expect(Fparser.calc('"foo"')).toEqual('foo');
+    });
+
+    it('allow spaces', function () {
```
this test fails.

bylexus commented on spec/specs/stringBasicSpec.js:
```diff
@@ -0,0 +1,96 @@
+        expect(() => Fparser.calc('2x', { x: 'foo' })).toThrowError(expectedError);
+    });
+
+    it('blocking math operator "+"', function () {
```
this test fails.

bylexus commented on spec/specs/stringBasicSpec.js:
```diff
@@ -0,0 +1,96 @@
+        expect(Fparser.calc('"   "')).toEqual('   ');
+    });
+
+    it('allow special chars', function () {
```
this test fails.

bylexus commented on spec/specs/stringBasicSpec.js:
```diff
@@ -0,0 +1,96 @@
+        });
+    });
+
+    it('blocking math operator "/"', function () {
```
this test fails.

bylexus commented on spec/specs/stringBasicSpec.js:
```diff
@@ -0,0 +1,96 @@
+        });
+    });
+
+    it('blocking math operator "*"', function () {
```
this test fails.

bylexus commented on spec/specs/stringBasicSpec.js:
```diff
@@ -0,0 +1,96 @@
+        expect(Fparser.calc('[myVar]', { myVar: 'foo' })).toEqual('foo');
+    });
+
+    it('support usage by custom function', function () {
```
this test fails.

bylexus reviewed (changes requested):
I am very sorry, but still, most of the tests fail when running `npm run test`. Maybe I am missing something?

Please correct either the failing code, or fix the tests (one of the two seems not correct).

LuigiPulcini commented:
Hi, @bylexus.

First of all, thank you so much for promptly reviewing my pull request. I am sorry I completely missed the fact that the tests would all fail with the library refactoring in Typescript.

I have now adjusted a few more details in the library and fixed all the tests accordingly. When I run `npm run test`, everything runs smoothly, now.

bylexus commented:
Hi @LuigiPulcini,

Thanks for your contribution, this is a valuable addition. I did some adaptions to your original pull request, mostly type definitions and some documentation.

For the time being, I merged it back to the `develop` branch, I want to implement some additional changes before merging it to the main and creating a version, namely:

- support for double- and single quote strings (now: only double quotes)
- make parser state names via enum, instead of error-prone strings

As soon as I added those additions, I will publish a new version.
